### PR TITLE
feat: add support for threads in federated rooms

### PR DESCRIPTION
**Pull request by MarcosSpessatto**
## Proposed changes (including videos or screenshots)

Jira: FED-87

### CODE REVIEW

File: apps/meteor/server/services/federation/application/room/input/RoomReceiverDto.ts [hunk 1]
```diff
@@ -330,6 +347,11 @@ export class FederationRoomReceiveExternalFileMessageDto extends ExternalMessage
 	messageBody: FederationFileMessageInputDto;
 
 	replyToEventId?: string;
+
+	thread?: {
+		rootEventId: string;
+		replyToEventId: string;
+	};
```
Comment (lmauromb) [inline apps/meteor/server/services/federation/application/room/input/RoomReceiverDto.ts:354]: should we define a thread type to use?

```typescript
type Thread {
  rootEventId: string;
  replyToEventId: string;
}
```

File: apps/meteor/tests/unit/server/federation/infrastructure/rocket-chat/converters/RocketTextParser.spec.ts [hunk 2]
```diff
@@ -545,6 +545,22 @@ describe('Federation - Infrastructure - Matrix - RocketTextParser', () => {
...
+			expect(
+				await toInternalQuoteMessageFormat({
+					homeServerDomain,
+					rawMessage,
+					formattedMessage: nested,
+					messageToReplyToUrl: 'http://localhost:3000/group/1?msg=2354543564',
+					senderExternalId: '@user:externalDomain.com',
+				}),
+			).to.be.equal('[ ](http://localhost:3000/group/1?msg=2354543564) hello nested quote');
+		});
```
Comment (lmauromb) [inline apps/meteor/tests/unit/server/federation/infrastructure/rocket-chat/converters/RocketTextParser.spec.ts:562]: nice 👍

**Review by lmauromb — APPROVED** [review]

File: apps/meteor/server/services/federation/infrastructure/matrix/Bridge.ts [hunk 3]
```diff
@@ -235,6 +235,162 @@ export class MatrixBridge implements IFederationBridge {
...
+						'is_falling_back': true,
+						'm.in_reply_to': {
+							event_id: relatesToEventId,
+						},
+					},
+				});
+			return messageId;
+		} catch (e) {
+			throw new Error('User is not part of the room.');
+		}
```
Comment (debdutdeb) [inline apps/meteor/server/services/federation/infrastructure/matrix/Bridge.ts:271]: this masks the actual error. in many scenarios can make troubleshooting a nightmare, might even impossible

File: apps/meteor/server/services/federation/infrastructure/matrix/Bridge.ts [hunk 4]
```diff
@@ -235,6 +235,162 @@ export class MatrixBridge implements IFederationBridge {
...
+					},
+				});
+
+			return messageId;
+		} catch (e: any) {
+			if (e.body?.includes('413') || e.body?.includes('M_TOO_LARGE')) {
+				throw new Error('File is too large');
+			}
+			return '';
+		}
```
Comment (debdutdeb) [inline apps/meteor/server/services/federation/infrastructure/matrix/Bridge.ts:348]: same here, can we log the actual error before returning?

File: apps/meteor/server/services/federation/infrastructure/matrix/Bridge.ts [hunk 5]
```diff
@@ -235,6 +235,162 @@ export class MatrixBridge implements IFederationBridge {
...
```
Comment (debdutdeb) [inline apps/meteor/server/services/federation/infrastructure/matrix/Bridge.ts:391]: same

**Review by debdutdeb — CHANGES_REQUESTED** [review]

**Pull request merged**